### PR TITLE
Model tools: Add various utility functions

### PR DESCRIPTION
**Pull request by cc7768**
In this pull request, I introduce an abstract type `AbstractUtility`. I also include 4 concrete utility functions and their derivatives: `LogUtility`, `CRRAUtility`, `CFEUtility`, and `EllipticalUtility` -- The first two are typically used in modeling consumption utility and second two for labor utility.

Additionally, the folder `modeltools` can be a placeholder for adding things like production functions etc... in the future.

Example

```julia
using QuantEcon

u = LogUtility()

# Evaluate log utility at c=1
@show u(1.0)

# Evaluate derivative of log utility at c=1
@show derivative(u, 1.0)
```

@sglyon

### CODE REVIEW

**Comment by natashawatkins** [comment]
This is a cool idea that could be used to replace some of the code in the optimal taxation lectures!

I guess you prefer `CRRAUtility` over `CRRAutility`. I should change the lecture if so...

**Comment by sglyon** [comment]
Hey @cc7768  let me know when you want a re view on this one.


I've seen a few changes come in since the PR opened os I've been waiting for your word to look closer

**Comment by cc7768** [comment]
I think this version is ready to review. There are 4 utility functions and their first derivative. There are also tests for each of them.

I will likely add some additional features (production function, aggregation functions, wage processes, etc...) in future PRs, but it is easier for me to do this in pieces than all at once.

File: src/modeltools/utility.jl [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+# Separable utility
+#
+
+# Consumption utility
+
+struct LogUtility <: AbstractUtility
+    ξ::Float64
+end
+
+LogUtility() = LogUtility(1.0)
```
Comment (sglyon) [inline src/modeltools/utility.jl]: Untested

File: src/modeltools/utility.jl [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    ifelse(c > 1e-10, u.ξ / c, 1e10)
+
+
+struct CRRAUtility <: AbstractUtility
+    γ::Float64
+    ξ::Float64
+
+    function CRRAUtility(γ, ξ=1.0)
+        if abs(γ - 1.0) < 1e-8
+            error("Your value for γ is very close to 1... Consider using LogUtility")
```
Comment (sglyon) [inline src/modeltools/utility.jl]: Untested.

File: src/modeltools/utility.jl [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+# Labor Utility
+
+struct CFEUtility <: AbstractUtility
+    ϕ::Float64
+    ξ::Float64
+
+    function CFEUtility(ϕ, ξ=1.0)
+        if abs(ϕ - 1.0) < 1e-8
+            error("Your value for ϕ is very close to 1... Consider using LogUtility")
```
Comment (sglyon) [inline src/modeltools/utility.jl]: Untested.

File: src/modeltools/utility.jl [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    ifelse(l > 1e-10, -u.ξ * l^(1.0/u.ϕ), 1e10)
+
+
+struct EllipticalUtility <: AbstractUtility
+    b::Float64
+    μ::Float64
+end
+
+# These defaults are pulled straight from Evans Phillips 2017
+EllipticalUtility(;b=0.5223, μ=2.2926) = EllipticalUtility(b, μ)
```
Comment (sglyon) [inline src/modeltools/utility.jl]: Untested

File: src/modeltools/utility.jl (hunk 1 above)
Comment (cc7768) [reply src/modeltools/utility.jl]: This should get tested now.

File: src/modeltools/utility.jl (hunk 2 above)
Comment (cc7768) [reply src/modeltools/utility.jl]: Should get tested now.

File: src/modeltools/utility.jl (hunk 3 above)
Comment (cc7768) [reply src/modeltools/utility.jl]: Should get tested now.

File: src/modeltools/utility.jl (hunk 4 above)
Comment (cc7768) [reply src/modeltools/utility.jl]: Should get tested now.

**Comment by sglyon** [comment]
Can you rebase off master to get the fixes @Shunsuke-Hori made to the calls to Optim?

That will get tests outside of this PR passing

**Comment by sglyon** [comment]
Actually, scratch that. 

I can see that these test pass b/c travis said the `pr` portion of the CI passed. When this is ready for another look let me know and we'll get it merged.

**Comment by cc7768** [comment]
I think this is ready for review.

Did you want me to rebase or is it ok?

**Pull request merged**